proposals/utils/checkers.py

```diff
@@ -472,14 +486,125 @@ def __init__(
+            fulfilled_children_slot = None
+            if self.study.has_children():
+                children_slots = [
+                    AttachmentSlot(
+                        self.study,
+                        kind=ConsentChildrenParents,
+                    ),
+                    AttachmentSlot(
+                        self.study,
+                        kind=ConsentChildrenNoParents,
+                    ),
+                ]
+
+                # if one of them has been fulfilled, this becomes the fulfilled slot
+                fulfilled_children_slot = self.at_least_one_fulfilled(children_slots)
+                for slot in children_slots:
+                    # if there is at least one, make the other one optional
+                    if fulfilled_children_slot and slot is not fulfilled_children_slot:
+                        slot.force_desiredness = desiredness.OPTIONAL
+                    self.stepper.add_slot(slot)
+
+            # Add a slot for a normal ConsentForm
+            self.stepper.add_slot(
+                AttachmentSlot(
+                    self.study,
+                    kind=ConsentForm,
+                    force_desiredness=(
+                        desiredness.REQUIRED
+                        if not fulfilled_children_slot
+                        else desiredness.OPTIONAL
+                    ),
+                )
+            )
+
         return []
 
+    def check_has_recordings(self):
+        """
+        A function that checks whether a study features audio or video
+        registration.
+        """
+        has_recordings = False
+        if self.study.get_observation:
+            # gather all AV observation_registraions
+            recordings_observation = observation_registration.objects.filter(
+                description__in=["audio recording", "video recording"]
+            )
+            # check if there is an overlap between these two QS's
+            if self.study.observation.registrations.all() & recordings_observation:
+                has_recordings = True
+        if self.study.get_sessions:
+            # gather all the tasks
+            all_tasks = Task.objects.filter(session__study=self.study)
+            # gather all AV task_registrations
+            recordings_sessions = task_registration.objects.filter(
+                description__in=["audio recording", "video recording"]
+            )
+            if all_tasks.filter(registrations__in=recordings_sessions):
+                has_recordings = True
+        return has_recordings
+
+    def at_least_one_fulfilled(self, slots):
+        """
+        Function which receives a list of slots and checks if at least one has
+        been fullfilled. It returns None, or the fullfiled slot
+        """
+        fullfilled_slot = None
+        for slot in slots:
+            # check if any of these slots have been fullfilled yet using match()
+            slot.match(exclude=[])
```

Comment:
This is somewhat abusive of the match function, because `slot.match()` modifies the internal state of the slot, setting its kind and attachment based on what it finds.

I think this currently does no tangible harm, but it's definitely something that can potentially go wrong in the future. I would like to request the following change:

* Modify `slot.match()` to just return the matched attachment or False if nothing is found. It should no longer change the state of the slot.
* Add a new `slot.match_and_set()` method that calls `match()`, and if it finds something, it sets `self.kind` and `self.attachment`.

Like so we can use `slot.match()` for use cases like in this checker, and use `match_and_set()` where we actually desire to change the state of the slot, such as in `stepper.add_slot()` and `stepper.attachment_slots()`.

Otherwise reusing the match method here is a great solution, I just don't like all these slots getting different attachments and kinds set for every pass of the checker.